_posts/2022-06-15-inline-assembly-memory-side-effects-bug.md

```diff
@@ -0,0 +1,156 @@
+of "medium".
+
+
+In the following example, the legacy code generation pipeline with enabled optimizer will
+remove the ``mstore`` and the function ``f`` will return zero:
+
+```solidity
+contract C {
+    function f() external pure returns (uint256 x) {
+        assembly {
+            mstore(0, 0x42)
+        }
+        assembly {
+            x := mload(0)
+        }
+    }
+}
+```
+
+However, if the same memory is either read again in the same inline assembly block
+or if the inline assembly block accesses any local Solidity variables, the bug is
+not present. Both is the case in the following example, so ``f`` will return
+``0x42`` as expected:
+
+```solidity
+contract C {
+    function f() external pure returns (uint256 x) {
+        assembly {
+            mstore(0, 0x42)
+            x := mload(0)
+        }
+    }
+}
+```
+
+In the following example, the first ``mstore`` will not be removed, since the written memory
+is read again by the ``return``. The second ``mstore`` on the other hand will be removed, since the memory
+it writes is never read again. In this case, this is a valid optimization and the example is not
+affected by the bug. Generally, any assembly block that terminates or does not have memory side-effects that
+need to be observed afterwards is unaffected.
+
+```solidity
+contract C {
+    function f() external {
+        assembly {
+            mstore(0, 0x42)  // This write will be kept, since the return below reads from the memory.
+            mstore(32, 0x21) // This will be removed, but that is valid since the memory is never read again.
+            return(0, 32)
+        }
+    }
+}
+```
+
+The most dangerous cases affected by the bug are those, in which fixed memory offsets are used to store intermediate
+values in one assembly block (e.g. in a helper function), which are then only used in a subsequent assemby block.
+
+```solidity
+contract C {
+    function callHelper() internal view {
+        assembly {
+            let ret := staticcall(gas(), address(), 0, 0, 0, 0)
+            if eq(ret, 0) {
+                revert(0, 0)
+            }
+            returndatacopy(0, 0, 128) // This will be removed due to the bug.
+        }
+    }
+    function f() external view returns(uint256 x) {
+        callHelper();
+        assembly {
+            // This consumes the memory write by the helper, which was incorrectly removed.
+            x := keccak256(0, 128)
+        }
+    }
+}
+```
+
+However, we found it to be uncommon for such patterns to occur: usually, an assembly block either reads from or writes
+to at least one local solidity variable, consumes the memory it writes directly, or takes over control until the end of
+the transaction. This is partly due to the fact that using fixed memory offsets for intermediate values between assembly
+blocks is in itself dangerous, since you need to ensure that the solidity code between the assembly block does not overwrite
```

Comment:
```suggestion
to at least one local Solidity variable, consumes the memory it writes directly, or takes over control until the end of
the transaction. This is partly due to the fact that using fixed memory offsets for intermediate values between assembly
blocks is in itself dangerous, since you need to ensure that the Solidity code between the assembly block does not overwrite
```